### PR TITLE
3825 web service endpoints

### PR DESCRIPTION
**Pull request by kjlippold**
[#3825]

This adds a data services section to the resource landing page which contains automatically generated GeoServer getCapabilities endpoints for public resources containing geographic feature or geographic raster content. Web Map Services are available for resources containing geographic feature and/or geographic raster content. Web Feature Services are available for resources containing geographic feature content, and Web Coverage Services are available for those containing geographic raster content. This also adds a right-click preview button to geographic feature and raster aggregations which will open a GeoServer generated preview of the data in a new tab.

### Pull Request Checklist: 
- [ ] Positive Test Case Written by Dev

- [ ] Automated Testing

- [ ] Sufficient User and Developer Documentation

- [ ] Passing Jenkins Build

- [ ] Peer Code review and approval

### Positive Test Case
1. [Enter positive test case here]

### CODE REVIEW

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-4023](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-4023)

File: hs_file_types/models/geofeature.py
Comment (cuahsimarko) [inline hs_file_types/models/geofeature.py:171]: What are the chances something unexpected will be in strings here, causing a server-side failure? Consider error handling, friendly user error and logging.

File: hs_composite_resource/models.py
Comment (cuahsimarko) [inline hs_composite_resource/models.py:5]: Are there considerations for different environments? Dev/UAT/QA/beta/prod? If so, make sure deployment staff understand the implications to Jenkins, settings, local_settings

File: hs_file_types/models/raster.py
Comment (cuahsimarko) [inline hs_file_types/models/raster.py:221]: consider error handling / logging

File: hydroshare/local_settings.template
Comment (cuahsimarko) [inline hydroshare/local_settings.template:184]: Please work with Dan / Martin and ensure they understand and update any necessary deployment scripts or manual steps

File: hs_composite_resource/models.py
Comment (cuahsimarko) [inline hs_composite_resource/models.py:639]: What happens when these services are down/broken? If appropriate, ensure a friendly error is passed to the enduser when clicking on Preview and something logged to help ops staff troubleshoot the hydroshare/hydroshare systems

**Comment by horsburgh** [comment]
A quick review of this functionality turned up the following issues:

1. This adds a new "Data Services" major heading to the Resource Landing Page. It's location right below the resource file content pane seems well chosen. However, I'm not sure it is fully separated from other elements on the page as it should be. The example testing resource is part of a collection. I looked at other resources in HydroShare that are part of a collection and the information about the collection should fall under the major heading of "References" (which is missing on the landing page for this testing resource) and should not be under the new "Data Services" major heading (see image below which shows the collection information without the "References" header):

image

2. The data service links take you to what I think is the "GetCapabilities" XML document served by GeoServer. While this is the correct URL to present, it's meaningless to a person who clicks it. We need some text that goes right under the new "Data Services" major header that says something like:

> The following geospatial data web services are available for this resource. The links below can be copied and pasted into Geographic Information Systems software programs to access the geospatial data included within this resource.

3. In the testing resource I viewed, the Citation is failing to generate. I'm assuming that this is unrelated to the new functionality that has been added, but want to make sure that merging this functionality in does not break something else.

image

File: theme/static/js/hs_file_browser.js
Comment (hyi) [inline theme/static/js/hs_file_browser.js:2028]: @kjlippold I don't seem to see any place where uiActionStates.preview.disabled is set to False and don't seem to be able to look at preview on the deployed test server, so I am wondering whether preview is working or not.

File: hs_file_types/models/raster.py
Comment (hyi) [inline hs_file_types/models/raster.py:250]: I see a lot of duplicate code in get_preview_data_url() in raster.py and geofeature.py. Suggest to move the common code into an utility function so it can be called by both functions in raster.py and geofeature.py.

**Comment by hyi** [comment]
@kjlippold I am done with code review and left a couple of comments.

**Comment by cuahsimarko** [comment]
Teams I've been on generally merged develop for large pull requests for a number of reasons, among them easier reviews as well as catching UI/or data subtleties. Interesting what Jeff noticed about the Citation, this should be investigated.

File: hs_composite_resource/models.py
Comment (sblack-usu) [inline hs_composite_resource/models.py:589]: All of this assumes the files are available on the geoserver and builds the url by convention.  It may make sense to set a flag on the resource object (within extra_data?) when the geoserver successfully retrieves the files.  This flag could be checked here so we have some assurance that the files are available on geoserver.  I'm open to other ideas and opinions on this.

File: hydroshare/local_settings.cutover
Comment (sblack-usu) [inline hydroshare/local_settings.cutover:196]: These config options look like they belong in settings.py, not local_settings.

Comment (sblack-usu) [reply hydroshare/local_settings.cutover:196]: The config names here are closely related to the HSWS* settings, so it would be good to have the names consistent.

Comment (kjlippold) [reply hydroshare/local_settings.cutover]: 867d2f9: Moved these settings to settings.py and changed all name patterns to HSWS*

File: hs_composite_resource/models.py
Comment (kjlippold) [reply hs_composite_resource/models.py:593]: We don't have anywhere in HydroShare to store information about the data services right now. For now, we've decided to build the URLs assuming the services are available.

File: hs_file_types/models/raster.py
Comment (kjlippold) [reply hs_file_types/models/raster.py:225]: 5e2905f: Moved duplicate code to a utility function.

File: theme/static/js/hs_file_browser.js
Comment (kjlippold) [reply theme/static/js/hs_file_browser.js:2028]: There's a GeoServer instance linked to prod and beta, but I haven't had a chance to set one up for dev-hs-6. By default, the links use the prod GeoServer, but since the resources on dev-hs-6 aren't registered on that GeoServer, you would expect to see a 404 page generated by Apache when you click the preview link. I'm working on getting another GeoServer set up, and I'll let you know once it's ready.

**Comment by kjlippold** [comment]
@horsburgh Responding to your comments above.

1. 2f43ab2: The References heading wasn't displaying for resources that belonged to a collection but didn't have sources or relations. That heading should appear now if the resource only belongs to a collection.

2. 8437ec4: I added a brief description under the heading. Let me know if there's anything else it should include, or if it should be formatted differently.

3. I wasn't able to reproduce this issue locally. I don't think it's related to this PR, but I'll keep looking.

File: hs_composite_resource/models.py
Comment (kjlippold) [reply hs_composite_resource/models.py:650]: f214a44: Added some basic error handling and logging. We don't really have a way right now for HydroShare to verify that each link is working as expected, but this should catch issues with how GeoServer is configured on HydroShare's end.

File: hs_file_types/models/raster.py
Comment (kjlippold) [reply hs_file_types/models/raster.py]: f214a44: Added some basic error handling and logging.

File: hs_file_types/models/geofeature.py
Comment (kjlippold) [reply hs_file_types/models/geofeature.py]: f214a44: Added some basic error handling and logging.

**Comment by martinseul** [comment]
In order to fully test the process on dev server I assume I would need to set up a new geoserver and link the HS instance to it. Is that process documented somewhere? I looked at using this repo https://github.com/CUAHSI-APPS/his_geoserver to set up the geoserver  but am unsure on the additional steps to get it all configured @kjlippold could you please document the steps.

File: hs_composite_resource/models.py
Comment (cuahsimarko) [reply hs_composite_resource/models.py:650]: If I understand, the URL will always appear in the UI, and the worst case scenario is the user clicks on it and gets a 404/500 or other page load error. Perhaps RENCI could add these URLs to what gets monitored in production. Not sure what systems we use to monitor such things. Just a thought. I believe the contacts for that would be Chris C. and Dan.

Comment (cuahsimarko) [reply hs_composite_resource/models.py:650]: If the URL would sometimes appear as an Error, then some message should be displayed in its place to email help@

File: hydroshare/local_settings.cutover
Comment (cuahsimarko) [reply hydroshare/local_settings.cutover]: Also, I'm not sure if local_settings.cutover is even used. @dan-palmer1 should investigate

**Comment by kjlippold** [comment]
@martinseul You’d also need to set up the web services manager app, but honestly even with the instructions it’s a bit complicated, so I’m working on simplifying the deployment of both of those components.

**Comment by martinseul** [comment]
Thanks @kjlippold that would be very helpful, where is the code for the web manager app currently?

**Comment by cuahsimarko** [comment]
@kjlippold I was wondering if you have finished with dev-hs-6 can you let me know?

**Comment by kjlippold** [comment]
@cuahsimarko I probably won’t have a chance to finish setting up the dev GeoServer instance until this weekend, so we’ll want to look at the PR once that’s ready, but I’m not actively using dev-hs-6 right now.

@martinseul The repo is at https://github.com/CUAHSI-APPS/hydroshare_web_services_manager. Again, it’s a little complex to set up, so I’m hoping to have a simpler setup sometime this weekend.

**Comment by cuahsimarko** [comment]
@kjlippold I will take back dev-hs-6 and if you need a server to deploy to next week, maybe beta will be fixed or Dan P. might have a server for you.

**Comment by hyi** [comment]
@cuahsimarko You can take back dev-hs-3 if that helps. Empty author fix testing has been done, so I don't need it any more.

File: hydroshare/settings.py
Comment (hyi) [inline hydroshare/settings.py:695]: @kjlippold Should the api token be kept as a secret? If so, that need to go into local_settings.

File: theme/templates/resource-landing-page/references.html
Comment (hyi) [inline theme/templates/resource-landing-page/references.html:4]: @kjlippold Why this ```belongs_to_collections``` condition is added here?

**Review by hyi — COMMENTED** [review]
@kjlippold I am done with another round of code review and left a couple of comments. Once they are addressed or responded, I'll approve this PR, which can then be merged pending functional testing approval from @martinseul

**Comment by hyi** [comment]
@kjlippold Just realized you need to resolve conflicts as well.

File: theme/templates/resource-landing-page/references.html
Comment (kjlippold) [reply theme/templates/resource-landing-page/references.html:4]: I added this condition in response to https://github.com/hydroshare/hydroshare/pull/4023#issuecomment-724189023. The 'References' major heading wasn't appearing in collections, so there wasn't a break between References and the preceding section. I'm not sure if this is intentional or not, but if it isn't it should probably be resolved separately.

File: hydroshare/settings.py
Comment (kjlippold) [reply hydroshare/settings.py:695]: Yes, it should be secret in production. I moved these settings from local_settings to settings in response to https://github.com/hydroshare/hydroshare/pull/4023#discussion_r520169829. I can move them back to local_settings if that's where they should be. We do want to keep these settings together though.

Comment (hyi) [reply hydroshare/settings.py:695]: OK. Then they should go to local_settings then. I assume the url and token you put here in settings are not real secrets but for testing only or for getting local development and a development auto-build working. If that is indeed the case, you can leave them here, and define the real values in local_settings with the same variable names, which will override whatever are defined in settings.

**Comment by hyi** [comment]
@kjlippold Looks like there is one test failure. After that is fixed and my local_settings/settings comment is responded, I can approve this PR which can be merged pending functional testing from @martinseul

File: hydroshare/local_settings.cutover
Comment (sblack-usu) [reply hydroshare/local_settings.cutover:195]: Looks like your commit moved all of the HSWS URLs over to settings.py.  I was referring only to the new settings.  This shouldn't be a problem though... it's pointing to a dev geoserver and not activated by default.

**Review by sblack-usu — APPROVED** [review]

**Comment by danames** [comment]
We a link to the help documentation added. We should put it after this sentence as follows: "The following Open Geospatial Consortium Web Services are available for this resource. The links below can be copied and pasted into Geographic Information Systems software programs to access geospatial data included within this resource. For more information on using these services visit the help pages."   The words "help pages" should link to https://help.hydroshare.org/apps. Where a "Using Data Services in Apps" link will be available.

**Pull request merged**
